src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/ModelExplorer.cs

```diff
@@ -0,0 +1,378 @@
+        {
+            _metadataProvider = metadataProvider;
+            Container = container;
+            Metadata = metadata;
+            _modelAccessor = modelAccessor;
+        }
+
+        /// <summary>
+        /// Creates a new <see cref="ModelExplorer"/>.
+        /// </summary>
+        /// <param name="metadataProvider">The <see cref="IModelMetadataProvider"/>.</param>
+        /// <param name="container">The container <see cref="ModelExplorer"/>.</param>
+        /// <param name="metadata">The <see cref="ModelMetadata"/>.</param>
+        /// <param name="model">The model object. May be <c>null</c>.</param>
+        public ModelExplorer(
+            [NotNull] IModelMetadataProvider metadataProvider,
+            [NotNull] ModelExplorer container,
+            [NotNull] ModelMetadata metadata,
+            object model)
+        {
+            _metadataProvider = metadataProvider;
+            Container = container;
+            Metadata = metadata;
+            _model = model;
+        }
+
+        /// <summary>
+        /// Gets the container <see cref="ModelExplorer"/>.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// The <see cref="Container"/> will most commonly be set as a result of calling
+        /// <see cref="GetExplorerForProperty(string)"/>. In this case, the returned <see cref="ModelExplorer"/> will
+        /// have it's <see cref="Container"/> set to the instance upon which <see cref="GetExplorerForProperty(string)"/>
+        /// was called.
+        /// </para>
+        /// <para>
+        /// This however is not a requirement. The <see cref="Container"/> is informational, and may not
+        /// represent a type that defines the property represented by <see cref="Metadata"/>. This can
+        /// occur when constructing a <see cref="ModelExplorer"/> based on evaluation of a complex
+        /// expression.
+        /// </para>
+        /// <para>
+        /// If calling code relies on a parent-child relationship between <see cref="ModelExplorer"/>
+        /// instances, then use <see cref="ModelMetadata.ContainerType"/> to validate this assumption.
+        /// </para>
+        /// </remarks>
+        public ModelExplorer Container { get; }
+
+        /// <summary>
+        /// Gets the <see cref="ModelMetadata"/>.
+        /// </summary>
+        public ModelMetadata Metadata { get; }
+
+        /// <summary>
+        /// Gets the model object.
+        /// </summary>
+        /// <remarks>
+        /// Retrieving the <see cref="Model"/> object will execute model accessor function if this
+        /// <see cref="ModelExplorer"/> was provided with one.
+        /// </remarks>
+        public object Model
+        {
+            get
+            {
+                if (_model == null && _modelAccessor != null)
+                {
+                    Debug.Assert(Container != null);
+                    _model = _modelAccessor(Container.Model);
```

Comment:
`Container.Model` is allowed to be null. In that case the accessor `m => 23` would be invoked as `null => 23`.  

`Container` is never allowed to be null in a deferred-evaluation scenario.